### PR TITLE
Add support for psr/cache 2 and 3 versions

### PR DESCRIPTION
**Pull request by rimvislt**

### CODE REVIEW

**Comment by Ocramius** [comment]
FWIW, let's drop PHP 7.4 as part of this? It has a shelf life of ~2 months left.

**Comment by rimvislt** [comment]
I was thinking if it is easier to drop php7 or try to load different files based on php version.

**Comment by rimvislt** [comment]
All test are passing minimum version is php8.0 now, let me know if anything else needed before this can be merged and released.

**Comment by rimvislt** [comment]
I am not sure about DCO not passing if I should do it or administrators. Let me know if I need to do anything otherwise everything else are passing now.

File: .laminas-ci.json
Comment (boesing) [inline .laminas-ci.json]: We can omit these here as with one of the latest versions of the matrix action I've changed the defaults of all versions to `false`

File: .laminas-ci.json
Comment (boesing) [inline .laminas-ci.json]: should be possible to have `*` here so we can reduce the amount of entries to 2 here. 👍🏼

File: src/Psr/CacheItemPool/CacheItem.php
Comment (boesing) [inline src/Psr/CacheItemPool/CacheItem.php]: Can we also add argument types so we are aligned with the interface?

File: src/Psr/CacheItemPool/CacheItem.php
Comment (boesing) [inline src/Psr/CacheItemPool/CacheItem.php:119]: Can we also add argument types so we are aligned with the interface?

File: src/Psr/CacheItemPool/CacheItem.php
Comment (boesing) [inline src/Psr/CacheItemPool/CacheItem.php:133]: Can we also add argument types so we are aligned with the interface?

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (boesing) [inline src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: Can we also add argument types so we are aligned with the interface?

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (boesing) [inline src/Psr/CacheItemPool/CacheItemPoolDecorator.php:151]: Can we also add argument types so we are aligned with the interface?

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (boesing) [inline src/Psr/CacheItemPool/CacheItemPoolDecorator.php:198]: Can we also add argument types so we are aligned with the interface?

File: test/Psr/CacheItemPool/CacheItemPoolDecoratorTest.php
Comment (boesing) [inline test/Psr/CacheItemPool/CacheItemPoolDecoratorTest.php:306]: do we have to provide these here?

**Review by boesing — CHANGES_REQUESTED** [review]
Nice! Thanks for working on this.
Just a few things but overall, good work 👍🏼

File: .laminas-ci.json
Comment (boesing) [reply .laminas-ci.json]: Ah, we should also add `*` for `php` since we now dropped PHP 7.4 support and thus can test all versions

File: test/Psr/CacheItemPool/CacheItemPoolDecoratorTest.php
Comment (rimvislt) [reply test/Psr/CacheItemPool/CacheItemPoolDecoratorTest.php:305]: Yes because psalm is complaining that can not understand data, since response is "iterable"

Comment (rimvislt) [reply test/Psr/CacheItemPool/CacheItemPoolDecoratorTest.php:306]: Yes, psalm does not work because original documentation inside interface gives just return type iterable and it is not enough. I tried to add it on implemented methods but psalm does not work with 2 return types defined (did not want to loose original doc type)

File: .laminas-ci.json
Comment (rimvislt) [reply .laminas-ci.json]: fixed

File: .laminas-ci.json
Comment (rimvislt) [reply .laminas-ci.json]: fixed

File: src/Psr/CacheItemPool/CacheItem.php
Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItem.php]: added (needed to drop "psr/cache": "^1.0  for it to work)

File: src/Psr/CacheItemPool/CacheItem.php
Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItem.php:119]: added (needed to drop "psr/cache": "^1.0  for it to work)

File: src/Psr/CacheItemPool/CacheItem.php
Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItem.php:133]: added (needed to drop "psr/cache": "^1.0  for it to work)

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: added (needed to drop "psr/cache": "^1.0  for it to work)

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:151]: added (needed to drop "psr/cache": "^1.0  for it to work)

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:198]: added (needed to drop "psr/cache": "^1.0  for it to work)

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (Ocramius) [inline src/Psr/CacheItemPool/CacheItemPoolDecorator.php]: Because this class is not `final`, the change in interface also implicitly becomes a BC break for people not depending on `psr/cache`.

@boesing that means new major version here too :S

File: phpcs.xml
Comment (Ocramius) [inline phpcs.xml:21]: If we roll a new major release, it would be fine to remove this exclusion.

As the current patch is presented, a new major version is indeed needed.

File: src/Psr/CacheItemPool/CacheItem.php
Comment (Ocramius) [reply src/Psr/CacheItemPool/CacheItem.php]: @boesing if you want the parameter types to match the interface, we will need a new major version here.

If we can keep `psr/cache:^1` and BC, then we can release a minor.

File: phpcs.xml
Comment (rimvislt) [reply phpcs.xml:21]: If we do major release we will need to add it into "doctrine/doctrine-module" it depends on v3 ->   "laminas/laminas-cache": "^3.1.2".

Let me know what is final decision on this and I will update it accordingly.

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (froschdesign) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php]: ~I think a major version is not needed at this point. The changes to the class `CacheItemPoolDecorator` can be undone, then a minor version will also work.~

Version 2 and 3 of the PSR-6 interfaces add type hints. (unfortunately they are not listed in the documentation)
Therefore, my comment here does not fit.

Comment (froschdesign) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: I think a major version is not needed at this point. The changes to the class `CacheItemPoolDecorator` can be undone, then a minor version will also work.

Comment (froschdesign) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: I think a major version is not needed at this point. The changes to the class `CacheItemPoolDecorator` can be undone, then a minor version will also work.

Comment (froschdesign) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: I think a major version is not needed at this point. The changes to the class `CacheItemPoolDecorator` can be undone, then a minor version will also work.

Comment (Ocramius) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php]: Tend to agree: new major can happen later, if we want :+1:

Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php]: Removed typed arguments and added back psr/cache v1 support.

File: phpcs.xml
Comment (Ocramius) [reply phpcs.xml:21]: Decided not to roll with a new major: resolving here.

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (Ocramius) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: Adding the return type here is still a BC break :| 

We're still in a "new major target" scenario, I fear...

Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: Most likely previously it was just return type in documentation but support for psr/cache v2 and v3 it needs to be added.

Comment (Ocramius) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: Yup: not blaming your changes, just saying that the lack of `final` has exposed us to BC issues here, since the dependency change leads to a transitive BC break.

Comment (froschdesign) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: I see no changes to support all three versions of PSR-6. Therefore it would be good to have plan before things are constantly being changed here.
@boesing, you as maintainer of this component, what is your plan for a new major version?

Comment (boesing) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: Oh, I've expected the decorators to be final since v3.
Seems that I forgot these...

I would be open for having a new major but only if we also add support for `psr/simple-cache` in its newer versions.

In the next major we can add `final` to the decorators as well. 

So we do have to drop v1 support of the psr packages as well in the next major? bummer...

File: test/Psr/CacheItemPool/CacheItemPoolDecoratorTest.php
Comment (boesing) [reply test/Psr/CacheItemPool/CacheItemPoolDecoratorTest.php:306]: Oh I see. Thats because they do not provide what is inside the iterable.
I wonder if its possible to add such stuff to the PSR packages...

File: src/Psr/CacheItemPool/CacheItemPoolDecorator.php
Comment (froschdesign) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: > I would be open for having a new major but only if we also add support for `psr/simple-cache` in its newer versions.

👍🏻

Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: So what is the plan with this one? Does it mean that v4 will happen in September?

Comment (boesing) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: I do not give an ETA but I will take care that this lands in v4.

Since that requires all supported adapters to support v4 as well plus the fact that there are some other things I've planned for v4, I dont think this will happen in september.

Comment (rimvislt) [reply src/Psr/CacheItemPool/CacheItemPoolDecorator.php:86]: Thank you for your response let me know how I can help with v4.

If you will create v4 branch  I will create new pr against it and will fix this code for it.

**Comment by rogervila** [comment]
Hi @rimvislt could you please fix the branch conflicts? Let me know if I can help on fixing this branch!

**Comment by rimvislt** [comment]
> 

I think above was decided that we will not use this pr and will be new version of it, so not sure if it is a point to fix it.

**Comment by boesing** [comment]
Yah, this PR provides a lot of changes which are not only related to `psr/cache` v3.

We will support `psr/cache` v2 & v3 and `psr/simple-cache` v2 & v3 with one of the next major versions as there is no way of introducing these versions without breaking BC.

----

I'd love to get some prepared PRs so I can simply merge them once we are creating the next major version(s). This will most likely happen after we created `laminas-servicemanager` v4 - even tho there is no official ETA for `servicemanager`, I am planning to create a release in the first quarter of 2023.

The changes in the PRs (please create dedicated PRs for both PSR packages) should contain:

- [ ] interface adaptions (i.e. added argument types along with return types - including `mixed`) within the delegator
- [ ] `composer.json` + `composer.lock` changes regarding the psr-package, this includes both `require` and `provides` keys

Thanks!

**Pull request closed without merging**
